protected/humhub/modules/content/components/ContentContainerActiveRecord.php

```diff
@@ -248,6 +249,40 @@ public function afterSave($insert, $changedAttributes)
         parent::afterSave($insert, $changedAttributes);
     }
 
+    /**
+     * @since 1.16
+     */
+    public function onSetPolymorphicRelationOwner(?Component $owner)
+    {
+        if (!$owner instanceof ContentContainer) {
+            return;
+        }
+
```

Comment:
Hmm, it is a special case and actually not an ideal implementation that in the `ContentContainer` case two records have the same GUID. 

So, I am unsure whether we should explicitly check the ContentContainer case here.

Reply:
Yes, the global GUID table would be the better apporach. But currently all the `guid` attributes are already harmonized between `ContentContainer` and `ContentContainerActiveRecord` models. Why do we need such a new handling here?

Reply:
`ContentContainerActiveRecord::afterSave($insert)` does this.


I would prefer to keep it that way. Maybe we could add a UUID option that `ContentContainerAR` has no GUID auto generation? 

Especially if it is planned to replace the code later, I would keep it as simple as possible.

Reply:
> `ContentContainerActiveRecord::afterSave($insert)` does this.

I see. Sorry, has been a while since I created the PR and have go get my head around it again.

So that's the line then:
https://github.com/humhub/humhub/pull/6587/files#diff-a2232aa8f7a9a3c12bcbc0443d6e59ec96f9a4d5e95e62b8498a0f9ad81a40d1L225
Which has been replaced in this PR with the functionality in `onSetPolymorphicRelationOwner()`

> I would prefer to keep it that way.

Ok, it can be reverted. It is still happening in `ContentContainerActiveRecord::afterSave()`, tough,. Just called in the following way:

- `ContentContainerActiveRecord::afterSave()`
  - `PolymorphicRelation::setPolymorphicRelation()`
    - `ContentContainerActiveRecord::onSetPolymorphicRelationOwner()`

> Maybe we could add a UUID option that `ContentContainerAR` has no GUID auto generation?

Yes, we could probably do that. IIRC it is also already possible. But the default is auto-assign. I'll look up what config we need here. But in both scenarios, the current as well as the new code, the GUID should be set already during validation. I realize that only now. Sorry.

> Especially if it is planned to replace the code later, I would keep it as simple as possible.

BTW,  the special case of ...
```php
if (!$owner instanceof ContentContainer) {
    return;
}
```
... will in practice never happen. It is _always_ a `ContentContainer`. I just added it to avoid a fatal error, since the `$cached` property of the PolymorphicRelationship instance could theoretically be any object and might not have a `guid` property...

Reply:
Ok, I understand. For me, a solution that is as simple and easy to understand as possible is important. Then comes elegance and flexibility. :-)

Hmm, since there is only one case with the identical GUID for `ContentContainer` & `ContentContainerAR`, I would prefer the simplest possible solution here.

e.g. 
- `ContentContainerAR and the UUID validator has an `allowEmpty` option
- `ContentContainerAR::afterSave` sets the GUID.

Reply:
I'm not sure this is a real-world example, but this pre-existing test leaves the $contentContainer without GUID. My patch would have made sure that also in this case, the correct GUID is assigned.

```php
  public function testGuid()
    {
        $user = User::findOne(['id' => 1]);

        // make sure we have a fresh ID and GUID
        $user->id = 9;
        $user->guid = UUID::v4();

        $contentContainer = new ContentContainer();
        $contentContainer->setPolymorphicRelation($user);

        $this->assertTrue($contentContainer->save());
        $this->assertEmpty($contentContainer->getErrors('guid'));
    }
```

Allowing the ContentContainer to have an empty guid does prevent the error from happening, but will result in a saved record without guid. Because even if you'd save the $user later, it would no longer update the guid in the $contentContainer, given it's not a new record to Yii.

But I guess this situation never happens in real.

Reply:
Ok, I understand. The situations (CC creation) are very limited.  Ideally, we refactor/fix the problem of duplicate GUID storage and thus also solve the cases.

Reply:
> Ok, I understand. The situations (CC creation) are very limited. Ideally, we refactor/fix the problem of duplicate GUID storage and thus also solve the cases.

As you have seen, I've created a PR that would address this:
- #6718 

However, in the meantime we need to solve this issue as the above mentioned test fails (also if it is in it's original) version, as it does test something that in real-world never happens.

We have two options:
- change the test to save the `$user` instead of the `$contentContainer`. Then the user's `afterSave()` would create a CC and add the guid. That would reflect what the real code is doing
- add the `GUID` behavior to `ContentContainer` (or add a `beforeSave` method) that fills in the `guid` upon save if there is none (i.e. if CCAR has not set it). But that could lead to a case where the CCAR and CC records have different `guid`s
- add a `setPolymorphicRelation()` method to `ContentContainer` that calls the behavior's `setPolymorphicRelation()` first and then, upon success, copies the `guid` from the CCAR. This would be a similar approach to what I had with the `onSetPolymorphicRelationOwner`-idea, but more straight forward. The function calls would be as follows
  - `ContentContainerActiveRecord::afterSave()`
    1. calling `ContentContainer::setPolymorphicRelation()`
       1. calling`PolymorphicRelation::setPolymorphicRelation()` (similar to a `parent::method()` call)
          - checking type and copying `id` and `class`
       2. copying guid if CC is new
    2. setting the other stuff it currently does  

The latest approach would allow to do stuff the test is doing, ie. 
1. create CC separately from CCAR, 
2. set the relation, 
3. save either CC or CCAR first

Would that work for you?

Reply:
It would look something like this: https://github.com/humhub/humhub/pull/6587/commits/e6e7c9bda8c5d5629cee525ef055f132df633afe

Reply:
Hmm, so if we always save the CCAR `$spaceOrUser->save()` in all places (Core, Rest API, User Importer), it would probably be easiest to just modify the test.  

Maybe it's also worth to add following method in CC

```php
    /**
     * @inheritDoc
     */
    public function save($runValidation = true, $attributeNames = null)
    {
        if ($this->isNewRecord && (empty($this->guid) || empty($this->class))) {
            throw new InvalidCallException(
                'Could not create standalone ContentContainer instance. ' .
                'Instance will be automatically created on ContentContainerActiveRecord::afterSave()');
        }

        return parent::save($runValidation, $attributeNames);
    }
```

Reply:
> Hmm, so if we always save the CCAR `$spaceOrUser->save()` in all places (Core, Rest API, User Importer), it would probably be easiest to just modify the test.

I don't know, if the test was the only place where CC was saved independently. However, since the `CCAR::afterSave()` creates a new instance of CC in any case for a new record, it would not make much sense and lead to an error, if CC was already created for this record.

Hence, yes I think it's safe to assume, that we always do that.

I have now updated the test.

> Maybe it's also worth to add following method in CC (if it's possible to catch `CC->save()`
> 
> ```
>     /**
>      * @inheritDoc
>      */
>     public function save($runValidation = true, $attributeNames = null)
>     {
>         if ($this->isNewRecord && (empty($this->pk) || empty($this->class))) {
>             throw new InvalidCallException(
>                 'Could not create standalone ContentContainer instance. ' .
>                 'Instance will be automatically created on ContentContainerActiveRecord::afterSave()');
>         }
> 
>         return parent::save($runValidation, $attributeNames);
>     }
> ```

Yes, that is a possiblity. Right now I've opted for validation instead. `class` was already required anyway, and `guid` is now, too.

If you prefer to have an error thrown, rather than validation fail, I'm happy to add your method. An exception is stronger than validation, since validation can be disabled and the result of `save()` ignored.

Whatever you prefer.

Reply:
Thanks I'm fine with the current solution.

If we save ourselves code/checks, we could also omit the new UUID validator from the `ContentContainer` model. In the future there should be a different solution here anyway?

Reply:
Cool.

Yes, with the [`global_id` table](https://github.com/humhub/humhub/pull/6718), the UUID validator would be used there, but not on those records implementing the proposed `GlobalActiveRecord` interface (indicating its a AR related to global_id)